src/internal/operations/manifests.go

```diff
@@ -127,52 +128,3 @@ func getManifestsAndMetadata(
 
 	return bb, collections, true, nil
 }
-
-// collectMetadata retrieves all metadata files associated with the manifest.
-func collectMetadata(
```

Comment:
I am not sure I fully agree with the argument. Here is what I think.

> Instead, this func shold remain where it is (we can assume we'll always need the basic set of metadata), and the provider/service-specific implementation should call a second, additional function that optionally extends the set of restoreCollections.

I don't think we should have the code first call this function to get the "basic" metadata and then call the method on `Controller` to get "additional metadata". This is introduces an unnecessary distinction between metadata belonging to services with only one underlying service and the rest(for example, in case of SP under group, there is no "basic" metadata).

> we have to duplicate those provider-agnostic parts

I don't think we are necessarily duplicating anything as the only implementation is in Controller and not in any of the specific services. It only calls out to providers that don't use the defaults like groups and asks it to return the set of metadata files we should use instead of the default set.

Reply:
Though the introduction of teams I'm increasingly bothered by the amount of code we've had to touch to enable or correct functionality purely by introducing the concept of the service.  The cognitive load is overwhelming, as is the amount of maintenance and the potential for bugs.  It's still early in the life of the repo and this problem will not scale well into the future.  "How do we do more with less work" is a philosophy we need to start adopting.

The fundamental question here is whether every provider should be _required_ to provide a metadata factory for the code to run successfully.  If we go that path, it's one more place where things don't work for the next addition until we add the support.  We don't often add providers, true.  But the direction still pushes us towards "do less with more".  Especially since this change removes code that is already doing more with less.

Anywhere that we can ensure the code works across changes without additional attention is good for our future maintenance and contribution, not to mention stability.

Reply:
> Anywhere that we can ensure the code works across changes without additional attention is good for our future maintenance and contribution, not to mention stability.

I might be missing something, but I feel like the current code retains the old functionality in that any new additional services that gets added do not have to do anything if they use the default metadata location. This is done by making use of [`groups.MetadataFiles`](https://github.com/alcionai/corso/blob/dd7617385855ffc5bd33c38cbad94028c1a397bf/src/internal/m365/backup.go#L209). This changes gives us a place where we can hook into and use non standard metadata in case the service deviates from it.

This changes also has minimal impact compared to introducing a custom struct and difference between "standard/basic" metadata and "additional" metadata.

---

I might also be misunderstanding your ask here. Are you suggesting that we could use the `collectMetadata` function to get the default set and add others on top by appending to the same array? One problem that I can see with it is that `sites` in `groups` do not have the standard set of metadata files. So it is useless (and possibly error prone) to try to get the default set of metadata first.

Reply:
> any new additional services

additional services, yes.  But it doesn't set us up well for the next provider.

Am I in the weeds a bit here?  Yes, absolutely, but even in the weeds the issue stands.  Perhaps a better way to frame my concern is from this perspective: if this code were its very first introduction- ie, the first time we started reading metadata before a backup- I'd have no problem moving forward with what you've written.  Without any other implementation to compare against, this is a perfectly adequate solution.  It'd get a pass.  Then, in the future, I'd be even happier to see a PR that refines this implementation by moving it out of the m365 package to treat it as a service-agnostic process.  Which would put things, more or less, at our current implementation.

It's not that your implementation has issues or significant impact.  It's that you're walking backwards from a goal post we've already reached.  This is why I like Ash's suggestion to have the provider return paths instead of collections.  It keeps us walking along the vector that got us past this goal post in the first place.  My prior suggestion involved redistributing metadata collection generation, which is also a step back from where we're at.  Is it also an acceptable solution?  Sure.  But not redistributing that code again is a better solution.

> Are you suggesting that we could use the collectMetadata function to get the default set and add others on top by appending to the same array? One problem that I can see with it is that sites in groups do not have the standard set of metadata files. So it is useless (and possibly error prone) to try to get the default set of metadata first.

This was my suggestion, yes.  Standard and non-standard files shouldn't pose any problems here because we're not passing files or file content around, we're using collections, and collections as a structure are resilient to differences in both the files and the directories they contain.  It is up to the service to describe the paths to those non-standard files, and it is up to that same service to consume those collections to retrieve and parse the files.  Those details are opaque to all other processes and consumers.

Reply:
>  Standard and non-standard files shouldn't pose any problems here because we're not passing files or file content around, we're using collections, and collections as a structure are resilient to differences in both the files and the directories they contain. It is up to the service to describe the paths to those non-standard files, and it is up to that same service to consume those collections to retrieve and parse the files. Those details are opaque to all other processes and consumers.

I think you misunderstood my concern here. The "standard" set of metadata includes `.../libraries/{previouspaths,delta}` for libraries, which would be returned in the collections, but we do not write a `delta` file at the top level for libraries.